### PR TITLE
Fix target manager CPU busyloop caused by bad done-channel handling.

### PR DESCRIPTION
**Pull request by juliusv**
Unfortunately this isn't nicely testable, as it's timing-dependent and
one would have to detect a stray goroutine doing a CPU busyloop...

Fixes https://github.com/prometheus/prometheus/issues/1114

### CODE REVIEW

File: retrieval/targetmanager.go [hunk 1]
```diff
@@ -113,7 +113,12 @@ type targetGroupUpdate struct {
 func (tm *TargetManager) Run() {
 	log.Info("Starting target manager...")
 
-	tm.done = make(chan struct{})
+	// It's important that we use a local copy of the "done" channel to listen
```
Comment (grobie) [inline retrieval/targetmanager.go]: It's not so much a copy than a reference, no? Alternatively you could also just explicitly pass the channel to he defer func.

Comment (juliusv) [inline retrieval/targetmanager.go]: Ok, I changed it to pass the channel directly to the functions below.

Comment (juliusv) [inline retrieval/targetmanager.go]: Ah, fracked up the push. Fixing.

Comment (juliusv) [inline retrieval/targetmanager.go]: Now.

**Comment by fabxc** [comment]
Might be worth checking whether this also fixes #1083.

**Comment by juliusv** [comment]
If you :+1: I'll merge and ask them to try.

**Comment by fabxc** [comment]
Edited in :+1: seconds after sending – apparently didn't propagate.

**Pull request merged**